### PR TITLE
Introducing 'debugging' and 'profiling' options in sonic build-infra

### PR DESCRIPTION
**Pull request by rodnymolina**
With these changes, user can now specify how debian packages will be built in regards to the presence of symbols and compiler-optimizations. I'm simply relying on DEB_BUILD_OPTIONS flags as indicated here: https://www.debian.org/doc/debian-policy/#debian-rules-and-deb-build-options

By default we will continue to build packages as is the case today.

```
Example after compilation with debugging OFF
============================================

rmolina@server04:~/wsp_gitli_Apr_30_2018/sonic-buildimage$ #find src/ -type f -executable -exec file '{}' \; | egrep "executable,.* stripped" > ~/executable_stripped_files_before.txt

rmolina@server04:~/wsp_gitli_Jun_2_2018/sonic-buildimage$ egrep "src\/sonic-*" ~/executable_stripped_files_before.txt | egrep ", stripped"
src/sonic-swss/debian/swss/usr/bin/swssconfig: ELF 64-bit LSB  executable, x86-64, version 1 (GNU/Linux), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=38a4858d3c9c479e89b0ec4a4d8acbf6f19795c7, stripped
src/sonic-swss/debian/swss/usr/bin/orchagent: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=33c611d547817f786942685f8495654b680cfe66, stripped
src/sonic-swss/debian/swss/usr/bin/teamsyncd: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=696f44f38f3c2e7acd88acdc51220f210c72ae99, stripped
src/sonic-swss/debian/swss/usr/bin/neighsyncd: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=d9bf4097548830a795488db9dbe26dbf4f0e194b, stripped
src/sonic-swss/debian/swss/usr/bin/intfmgrd: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=09ef06341bc7748d62ad926479373726a191acf0, stripped
src/sonic-swss/debian/swss/usr/bin/swssplayer: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=685c29e90bf4a080ca104da3f11ee4bc8e21dbc1, stripped
src/sonic-swss/debian/swss/usr/bin/portsyncd: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=5089a919e03fb842e10e23ef4be77d4add215e32, stripped
src/sonic-swss/debian/swss/usr/bin/fpmsyncd: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=4cb4cdebda61921c3367584e965bb5d230466b42, stripped
src/sonic-swss/debian/swss/usr/bin/routeresync: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=5a161cfa375e0bba789068e7c661e160b18b8615, stripped
src/sonic-swss/debian/swss/usr/bin/vlanmgrd: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=a4c0b8b6ee8984da6827aa15fc305257ca868326, stripped
src/sonic-swss/debian/swss/usr/bin/buffermgrd: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=607d8d2cef930ccf89d649c895000c09dda59139, stripped
src/sonic-swss/debian/swss/usr/bin/intfsyncd: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=08d323095731fdae0667dd9b155090ecf8ef4804, stripped
src/sonic-linux-kernel/linux-3.16.51/debian/build/build_amd64_none_amd64/arch/x86/boot/compressed/vmlinux.bin: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), statically linked, BuildID[sha1]=613f30030d61a0b7e46f6631c76489533ffe5d81, stripped
src/sonic-sairedis/debian/syncd-rpc/usr/bin/syncd: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=05b889a134292f7f200352ec90742c72414c880a, stripped
src/sonic-sairedis/debian/syncd-rpc/usr/bin/syncd_request_shutdown: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=aa6dd0376c90c8c9fd4da8ba55c464d0bae4e102, stripped
src/sonic-sairedis/debian/syncd-rpc/usr/bin/saiplayer: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=174554fd3857460d09e3207d65b792f786367c93, stripped
src/sonic-sairedis/debian/syncd-rpc/usr/bin/saidump: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=36588b3870e7b579fba99101433c397a8f32cd9f, stripped
src/sonic-sairedis/debian/syncd-rpc/usr/bin/saisdkdump: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=4eaa3369298343a69729228f7aa81a1b6e9c2843, stripped
src/sonic-swss-common/debian/libswsscommon/usr/bin/swssloglevel: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.32, BuildID[sha1]=82aeca06ad2c54384fd98ce79c6c98d12f7bfec9, stripped

Example after compilation with debugging ON
============================================

rmolina@server04:~/wsp_gitli_Jun_2_2018/sonic-buildimage$ find src/ -type f -executable -exec file '{}' \; | egrep "executable,.* stripped" > ~/executable_stripped_files_after.txt

rmolina@server04:~/wsp_gitli_Jun_2_2018/sonic-buildimage$ egrep "src\/sonic-*" ~/executable_stripped_files_after.txt | egrep ", stripped"
src/sonic-linux-kernel/linux-3.16.51/debian/build/build_amd64_none_amd64/arch/x86/boot/compressed/vmlinux.bin: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), statically linked,
```

### CODE REVIEW

File: slave.mk [hunk 1]
```diff
@@ -86,6 +86,14 @@ else
...
 endif
 
+ifeq ($(SONIC_DEBUGGING_ON),y)
+DEB_BUILD_OPTIONS_GENERIC := "nostrip"
+endif
+
+ifeq ($(SONIC_PROFILING_ON),y)
+DEB_BUILD_OPTIONS_GENERIC := "nostrip noopt"
+endif
+
```
[the remark this answers is not in the document]
Comment (rodnymolina) [reply slave.mk]: Why would you want to do that? 'profiling' option is a superset of 'debugging' one, so i can't think about a use-case to enable both. Let me know if i didn't understand your question correctly.

[the remark this answers is not in the document]
Comment (jleveque) [reply slave.mk]: I think the concern is that the 'profiling' option is a superset of 'debugging, yet there is nothing *preventing* someone from enabling both. Maybe just enhance the comment?

[the remark this answers is not in the document]
Comment (rodnymolina) [reply slave.mk]: Got it. I'll clarify that in rules/config. Basically, if you enable both options, then only the 'superset' one (profiling) will be activated, as this one appears in slave.mk _after_ 'debugging' option -- i placed them in this order on purpose.

[the remark this answers is not in the document]
Comment (rodnymolina) [reply slave.mk]: Btw, just as an indication of how much larger sonic-broadcom.bin would get with 'debugging' option turned on: we are seeing an increment from the current ~420MB to ~495MB, which in our case definitely pays off to make our life easier at t-shooting time -- in fact we left 'debugging' option enabled by default.

File: rules/config [hunk 2]
```diff
@@ -56,5 +56,11 @@ SONIC_ROUTING_STACK = quagga
 # Enable Origanization Extensions - Specific to the deployment scenarios of the Organization
 ENABLE_ORGANIZATION_EXTENSIONS = y
 
+# Debugging option allows sonic debian packages to get built including symbols
+# information. Profiling option, disables compiler optimizations (-O0) as well
+# as includes symbols information too.
+#SONIC_DEBUGGING_ON = y
```
Comment (jleveque) [inline rules/config]: I feel as though this is a bit confusing, as we also have the `SONIC_CONFIG_DEBUG` option at line 43, which installs packages needed for debug.

Comment (rodnymolina) [reply rules/config]: What about adding a better a description for SONIC_CONFIG_DEBUG option too? i can do something like: s/install debug packages/Install linux debugging tools/

Comment (rodnymolina) [reply rules/config]: Or we can also change SONIC_DEBUGGING_ON to something else. Let me know if have a better  name for this.

Comment (jleveque) [reply rules/config]: Maybe change `SONIC_CONFIG_DEBUG` to something like `SONIC_CONFIG_DEBUG_TOOLS`, `SONIC_CONFIG_INSTALL_DEBUG_TOOLS` or `SONIC_INSTALL_DEBUG_TOOLS`?

Comment (rodnymolina) [reply rules/config]: Will go ahead with this one: SONIC_INSTALL_DEBUG_TOOLS

File: rules/config [hunk 3]
```diff
@@ -56,5 +56,13 @@ SONIC_ROUTING_STACK = quagga
 # Enable Origanization Extensions - Specific to the deployment scenarios of the Organization
 ENABLE_ORGANIZATION_EXTENSIONS = y
 
+# Debugging option allows sonic debian packages to get built including symbols
+# information. Profiling option, disables compiler optimizations (-O0) as well
+# as includes symbols information. Being 'profiling' option a superset of
```
Comment (jleveque) [inline rules/config]: Being 'profiling' option **is** a superset... [Missing "is"]

**Comment by jleveque** [comment]
Retest this please.

**Pull request merged**